File: ExtremeParser/Types/DefinesStruct.vb
Imports System.Text.RegularExpressions

Public Structure DefinesStruct

    'No properties for speed.
    Public DefineName As String
    Public DefineValue As String

    Public Sub New(defName As String, defValue As String)
        DefineName = defName
        DefineValue = defValue
    End Sub

    Public Shared Operator =(ByVal first As DefinesStruct, second As DefinesStruct)
        If first.DefineName = second.DefineName And first.DefineValue = second.DefineValue Then
            Return True
        End If
        Return False
    End Operator
    Public Shared Operator <>(ByVal first As DefinesStruct, second As DefinesStruct)
        If first = second Then
            Return False
        Else
            Return True
        End If
    End Operator
End Structure

File: ExtremeParser/Types/DefinesStruct.cs
namespace ExtremeParser.Types
{
    public struct DefinesStruct
    {
        public bool Equals(DefinesStruct other)
        {
            return string.Equals(DefineName, other.DefineName) && string.Equals(DefineValue, other.DefineValue);
        }

        public override bool Equals(object obj)
        {
            if (ReferenceEquals(null, obj)) return false;
            return obj is DefinesStruct && Equals((DefinesStruct) obj);
        }

        public override int GetHashCode()
        {
            unchecked
            {
                return ((DefineName != null ? DefineName.GetHashCode() : 0) * 397) ^ (DefineValue != null ? DefineValue.GetHashCode() : 0);
            }
        }

        //No properties for speed.
        public string DefineName;
        public string DefineValue;

        public DefinesStruct(string defName, string defValue)
        {
            DefineName = defName;
            DefineValue = defValue;
        }

        public static object operator ==(DefinesStruct first, DefinesStruct second)
        {
            return first.DefineName == second.DefineName && first.DefineValue == second.DefineValue;
        }

        public static object operator !=(DefinesStruct first, DefinesStruct second)
        {
            return first.DefineName != second.DefineName || first.DefineValue != second.DefineValue;
        }
    }
}